src/Tuura/Concept/FSM/Translation.hs

```diff
@@ -81,46 +81,55 @@ translateFSM circuitName ctype signs = do
+              sortedCause = concatMap handleArcs (groupSortOn snd (arcLists allCause))
               initialState = getInitialState circuit signs
               allArcs = createAllArcs sortedCause
               reachables = findReachables allArcs initialState
-              invariants = map encToInt (concatMap (\i -> getInvariantStates i signs) (invariant circuit))
+              invariants = concatMap (\i -> getInvariantStates i signs) (invariant circuit)
+              invariantNos = map encToInt invariants
               reachableArcs = removeUnreachables allArcs reachables
               inputSigns = filter ((==Input) . interface circuit) signs
               outputSigns = filter ((==Output) . interface circuit) signs
               internalSigns = filter ((==Internal) . interface circuit) signs
-              reachableInvariants = filter (`elem` reachables) invariants
-              unreachables = ([0..2^(length signs) - 1] \\ invariants) \\ reachables
-          if (reachableInvariants /= [])
-              then reachableInvariantStateError reachableInvariants
-            else do
-              let reachReport = genReachReport unreachables
-              genFSM inputSigns outputSigns internalSigns (map show reachableArcs) (show initialState) reachReport
-      Invalid errs ->
-          "Error. \n" ++ addErrors errs
+              reachableInvariants = filter (\i -> (encToInt i) `elem` reachables) invariants
+              unreachables = ([0..2^(length signs) - 1] \\ invariantNos) \\ reachables
+          case validateFSM signs circuit reachableInvariants (invariant circuit) of
+              Valid -> do
+                let reachReport = genReachReport unreachables
+                genFSM inputSigns outputSigns internalSigns (map show reachableArcs) (show initialState) reachReport
+              Invalid errs -> addErrors errs
+
+      Invalid errs -> addErrors errs
 
 getInitialState :: CircuitConcept a -> [a] -> Int
 getInitialState circuit signs = encToInt state
   where
-    state = map (fromBool  . getDefined . initial circuit ) signs
+    state = map (fromBool  . getDefined . initial circuit) signs
 
 fromBool :: Bool -> Tristate
 fromBool x = if x then triTrue else triFalse
 
-addConsistency :: Ord a => [([Transition a], Transition a)] -> [a] -> [([Transition a], Transition a)]
-addConsistency allArcs signs = nubOrd (allArcs ++ concatMap (\s -> [([rise s], fall s), ([fall s], rise s)]) signs)
+addConsistency :: Ord a => [Causality (Transition a)] -> [a] -> [Causality (Transition a)]
+addConsistency allArcs signs = nubOrd (allArcs ++ concatMap (\s -> [AndCausality (rise s, fall s), AndCausality (fall s, rise s)]) signs)
 
 handleArcs :: [([Transition a], Transition a)] -> [([Transition a], Transition a)]
-handleArcs arcLists = result
+handleArcs xs = map (\m -> (m, effect)) transCauses
         where
-            effect = snd (head arcLists)
-            effectCauses = map fst arcLists
+            effect = snd (head xs)
```

Comment:
This will fail if `xs == []` -- is this expected? If no, let's fix this. If there is an assumption that the input list is non-empty, let's at least add a comment about this.